packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -575,7 +598,8 @@ abstract class TestRunner {
 
   private doMutations(mutations: Mutation[]): Promise<void> {
     const userCallback = new Deferred<void>();
-    this.outstandingWrites.push({ mutations, userCallback });
+    this.outstandingMutations.push(mutations);
+    this.outstandingCallbacks.push(userCallback);
```

Comment:
I originally wrote this:

> _So whenever we push to outstandingMutations we also push to outstandingCallbacks, but when we poll() from outstandingMutations we only _sometimes_ shift() from outstandingCallbacks (based on whether the step had expectUserCallback: true or not).  So it seems like these lists are going to end up out-of-sync (after a expectUserCallback: false step, subsequent steps will end up waiting on the wrong user callback).  But I'd expect to cause tests to fail...  so maybe I'm confused?_

But after staring at this for several minutes I figured out what's going on.  outstandingMutations is shared across clients but outstandingCallbacks is per-client, and outstandingCallbacks is sometimes consumed on its own without outstandingMutations (by doDrainQueue). And so everything should balance out.

This is a bit confusing / non-obvious. Maybe rename outstandingMutations to globalOutstandingMutations or crossClientOutstandingMutations or something? And can you comment outstandingMutations and outstandingCallbacks and their relationship to each other and the fact that outstandingMutations is shared across clients?

Reply:
As per https://github.com/firebase/firebase-js-sdk/pull/678, `outstandingCallbacks` is no longer a thing. `outstandingMutations` is now also `sharedWrites` and hopefully clears up some of the confusion.